### PR TITLE
[ROCKETMQ-166] onException callback may capture compressed message body

### PR DESCRIPTION
**Pull request by lizhanhui**
If message body size exceeds specified threshold, client would try to compress the message body.
Here there are two issues: 1) current implementation changes message body directly, which is not a good practice; 2) if asynchronous send method is employed to deliver message, when onException is invoked, the callback may capture the compressed message body before the finally block restores it.

### CODE REVIEW

**Comment by lizhanhui** [comment]
Comment to bump up this PR.

**Comment by vongosling** [comment]
@shroman @stevenschew

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java [hunk 1]
```diff
@@ -306,6 +310,19 @@ public SendResult sendMessage(//
...
     ) throws RemotingException, MQBrokerException, InterruptedException {
+
+        byte[] msgBody = msg.getBody();
+        Integer sysFlag = requestHeader.getSysFlag();
+        if (null != sysFlag && ((sysFlag & MessageSysFlag.COMPRESSED_FLAG) == MessageSysFlag.COMPRESSED_FLAG)) {
+            try {
+                msgBody = UtilAll.compress(msgBody, zipCompressLevel);
+            } catch (IOException e) {
+                requestHeader.setSysFlag(MessageSysFlag.clearCompressedFlag(sysFlag));
+                log.error("tryToCompressMessage exception", e);
```
Comment (shroman) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: What is the reason to have ERROR and WARN here?

Comment (lizhanhui) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: You mean why we log error on failing to compress message body?

**Comment by lizhanhui** [comment]
@stevenschew Any idea?

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java (hunk 1 above)
Comment (shroman) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: I just wonder why both WARN and ERROR are needed.

Comment (lizhanhui) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: Got your point.  

1. The code is moved here, and same log level should be applied IMO. @vongosling @vintagewang  any particular reason behind this?

**Comment by vongosling** [comment]
@lizhanhui I will close the pr, if you happened to the same question, please let me know.

**Pull request closed without merging**